File: code/dynamic_programming/src/weighted_job_scheduling/weighted_job_scheduling.cpp
/*
 * Part of Cosmos by OpenGenus Foundation
 * C++ program for weighted job scheduling using Dynamic Programming and Binary Search
 */

#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

// A job has start time, finish time and profit.
//index  is used to store its given position as we will going to do sorting
struct Job
{
    int start, finish, profit, index;
};


// A utility function that is used for sorting events
// according to finish time
bool jobComparator(Job s1, Job s2)
{
    return s1.finish < s2.finish;
}

// A Binary Search based function to find the latest job
// (before current job) that doesn't conflict with current
// job. "index" is index of the current job. This function
// returns -1 if all jobs before index conflict with it. The
// array jobs[] is sorted in increasing order of finish time
int latestNonConflict(const vector<Job> &jobs, int index)
{
    // Initialize 'lo' and 'hi' for Binary Search
    int lo = 0, hi = index - 1;

    // Perform binary Search iteratively
    while (lo <= hi)
    {
        int mid = (lo + hi) / 2;
        if (jobs[mid].finish <= jobs[index].start)
        {
            if (jobs[mid + 1].finish <= jobs[index].start)
                lo = mid + 1;
            else
                return mid;
        }
        else
            hi = mid - 1;
    }

    return -11;             //-11 just a flag value but kept as same to that given to index 0 see: MaxProfit func
                            //so that backtracking can be done
}

// The main function that finds the subset of jobs
// associated with maximum profit such that no two
// jobs in the subset overlap.
void MaxProfit(vector<Job> &arr)
{

    int n = arr.size();
    // Sort jobs according to finish time
    sort(arr.begin(), arr.end(), jobComparator);           //reason for using index

    // Create an array to store solutions of subproblems.
    // DP[i] stores the Jobs involved and their total profit
    // till arr[i] (including arr[i])
    vector<int> DP(n + 2);

    // initialize DP[0] to arr[0]
    DP[0] = arr[0].profit;

    //contain what is the latest non conflicting job with the current job
    vector<int> nConflict;

    //no job non-conflicting with one job
    nConflict.push_back(-11);             //-11 is just a flag value

    for (int i = 1; i < n; i++)
        nConflict.push_back(latestNonConflict(arr, i));

    // Fill entries in DP[] using recursive property
    //recurrence
    for (int i = 1; i < n; i++)
    {
        int val = nConflict[i];
        DP[i] = max(DP[i - 1], DP[val] + arr[i].profit);
    }

    //finding jobs included in max profit
    int i = n - 1;
    cout << "Optimal jobs are -  ";
    while (true)
    {

        //terminating condition
        if (i < 0)                  //basically checking case of if it is equal to index of any job --(-11)
            break;

        //DP[i] will be one of the two values from which it is derived(shown above)

        if (i == 0)
        {
            cout << arr[i].index + 1;
            break;
        }
        else if (DP[i] == DP[i - 1])
            i--;

        else
        {
            cout << arr[i].index + 1 << " ";
            i = nConflict[i];
        }
    }

    cout << "\nMaximum Profit = " << DP[n - 1] << "\n";
}

int main()
{
    //example test case
    vector<Job> arr(4);
    arr[0] = {3, 5, 20, 0};
    arr[1] = {1, 2, 50, 1};
    arr[2] = {6, 19, 100, 2};
    arr[3] = {2, 100, 200, 3};

    MaxProfit(arr);

    return 0;
}

File: code/dynamic_programming/src/weighted_job_scheduling/weighted_job_scheduling.py
# Python3 program for weighted job scheduling using
# Naive Recursive Method

# Importing the following module to sort array
# based on our custom comparison function
from functools import cmp_to_key

# A job has start time, finish time and profit
class Job:
	
	def __init__(self, start, finish, profit):
		
		self.start = start
		self.finish = finish
		self.profit = profit

# A utility function that is used for
# sorting events according to finish time
def jobComparataor(s1, s2):
	
	return s1.finish < s2.finish

# Find the latest job (in sorted array) that
# doesn't conflict with the job[i]. If there
# is no compatible job, then it returns -1
def latestNonConflict(arr, i):
	
	for j in range(i - 1, -1, -1):
		if arr[j].finish <= arr[i - 1].start:
			return j
			
	return -1

# A recursive function that returns the
# maximum possible profit from given
# array of jobs. The array of jobs must
# be sorted according to finish time
def findMaxProfitRec(arr, n):
	
	# Base case
	if n == 1:
		return arr[n - 1].profit

	# Find profit when current job is included
	inclProf = arr[n - 1].profit
	i = latestNonConflict(arr, n)
	
	if i != -1:
		inclProf += findMaxProfitRec(arr, i + 1)

	# Find profit when current job is excluded
	exclProf = findMaxProfitRec(arr, n - 1)
	return max(inclProf, exclProf)

# The main function that returns the maximum
# possible profit from given array of jobs
def findMaxProfit(arr, n):
	
	# Sort jobs according to finish time
	arr = sorted(arr, key = cmp_to_key(jobComparataor))
	return findMaxProfitRec(arr, n)

# Driver code
values = [ (3, 10, 20), (1, 2, 50),
		(6, 19, 100), (2, 100, 200) ]
arr = []
for i in values:
	arr.append(Job(i[0], i[1], i[2]))
	
n = len(arr)

print("The optimal profit is", findMaxProfit(arr, n))

# This code is code contributed by Kevin Joshi
